python/pyspark/sql/types.py

```diff
@@ -1629,35 +1629,121 @@ def to_arrow_type(dt):
+                # TODO: handle nested timestamps, such as ArrayType(TimestampType())?
+                if is_datetime64tz_dtype(series.dtype):
+                    pdf[column] = series.dt.tz_convert(tz).dt.tz_localize(None)
+                elif is_datetime64_dtype(series.dtype) and timezone is not None:
+                    # `series.dt.tz_localize('tzlocal()')` doesn't work properly when including NaT.
+                    pdf[column] = series.apply(
+                        lambda ts: ts.tz_localize('tzlocal()').tz_convert(tz).tz_localize(None)
+                        if ts is not pd.NaT else pd.NaT)
+    except ImportError:
+        from pandas.core.common import is_datetime64_dtype
+        from pandas.tslib import _dateutil_tzlocal
+        tzlocal = _dateutil_tzlocal()
+        tz = timezone or tzlocal
+        for column, series in pdf.iteritems():
+            if type(schema[str(column)].dataType) == TimestampType:
+                # TODO: handle nested timestamps, such as ArrayType(TimestampType())?
+                if not is_datetime64_dtype(series.dtype):
+                    # `series.dt.tz_convert(tzlocal).dt.tz_localize(None)` doesn't work properly.
+                    pdf[column] = pd.Series([ts.tz_convert(tz).tz_localize(None)
+                                             if ts is not pd.NaT else pd.NaT for ts in series])
+                elif is_datetime64_dtype(series.dtype) and timezone is not None:
+                    # `series.dt.tz_localize(tzlocal)` doesn't work properly.
+                    pdf[column] = pd.Series(
+                        [ts.tz_localize(tzlocal).tz_convert(tz).tz_localize(None)
+                         if ts is not pd.NaT else pd.NaT for ts in series])
     return pdf
 
 
-def _check_series_convert_timestamps_internal(s):
+def _check_series_convert_timestamps_internal(s, timezone):
     """
     Convert a tz-naive timestamp in local tz to UTC normalized for Spark internal storage
     :param s: a pandas.Series
     :return pandas.Series where if it is a timestamp, has been UTC normalized without a time zone
     """
-    from pandas.api.types import is_datetime64_dtype, is_datetime64tz_dtype
-    # TODO: handle nested timestamps, such as ArrayType(TimestampType())?
-    if is_datetime64_dtype(s.dtype):
-        return s.dt.tz_localize('tzlocal()').dt.tz_convert('UTC')
-    elif is_datetime64tz_dtype(s.dtype):
-        return s.dt.tz_convert('UTC')
-    else:
-        return s
+    try:
+        from pandas.api.types import is_datetime64tz_dtype, is_datetime64_dtype
+        # TODO: handle nested timestamps, such as ArrayType(TimestampType())?
+        if is_datetime64_dtype(s.dtype):
+            tz = timezone or 'tzlocal()'
+            return s.dt.tz_localize(tz).dt.tz_convert('UTC')
+        elif is_datetime64tz_dtype(s.dtype):
+            return s.dt.tz_convert('UTC')
+        else:
+            return s
+    except ImportError:
```

Comment:
I think we should bump up pandas version if we can't find a workaround.

Reply:
Sure, let me look into it a little more and summarize what version we can support.